go/tools/releaser/upgradedep.go

```diff
@@ -0,0 +1,518 @@
+	// in the directory. If there is no such directive, we apply the old patch
+	// using "patch". In either case, we'll generate a new patch with "diff".
+	// We'll scrub the timestamps to avoid excessive diffs in the PR that
+	// updates dependencies.
+	rootDir, err := repoRoot()
+	if err != nil {
+		return err
+	}
+	if attrs["patches"] != nil {
+		if err != nil {
+			return err
+		}
+		patchDir := filepath.Join(workDir, name, "a")
+		if err := extractArchive(archiveFile, path.Base(ghURL), patchDir, stripPrefix); err != nil {
+			return err
+		}
+
+		patchesList, ok := (*attrs["patches"]).(*bzl.ListExpr)
+		if !ok {
+			return fmt.Errorf("\"patches\" attribute is not a list")
+		}
+		for patchIndex, patchLabelExpr := range patchesList.List {
+			patchLabel, ok := patchLabelExpr.(*bzl.StringExpr)
+			if !ok {
+				return fmt.Errorf("not all patches are string literals")
+			}
+			if !strings.HasPrefix(patchLabel.Value, "@io_bazel_rules_go//third_party:") {
+				return fmt.Errorf("patch does not start with '@io_bazel_rules_go//third_party:': %s", patchLabel)
+			}
+			patchName := patchLabel.Value[len("@io_bazel_rules_go//third_party:"):]
+			patchPath := filepath.Join(rootDir, "third_party", patchName)
+			prevDir := filepath.Join(workDir, name, string('a'+patchIndex))
+			patchDir := filepath.Join(workDir, name, string('a'+patchIndex+1))
+			var patchCmd []string
+			for _, c := range patchLabel.Comment().Before {
+				words := strings.Fields(strings.TrimPrefix(c.Token, "#"))
+				if len(words) > 0 && words[0] == "releaser:patch-cmd" {
+					patchCmd = words[1:]
+					break
+				}
+			}
+
+			if err := copyDir(patchDir, prevDir); err != nil {
+				return err
+			}
+			if patchCmd == nil {
+				if err := runForError(ctx, patchDir, "patch", "-Np1", "-i", patchPath); err != nil {
+					return err
+				}
+			} else {
+				if err := runForError(ctx, patchDir, patchCmd[0], patchCmd[1:]...); err != nil {
+					return err
+				}
+			}
+			patch, _ := runForOutput(ctx, filepath.Join(workDir, name), "diff", "-urN", string('a'+patchIndex), string('a'+patchIndex+1))
+			patch = sanitizePatch(patch)
+			if err := os.WriteFile(patchPath, patch, 0666); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Update the attributes.
+	*attrs["sha256"] = &bzl.StringExpr{Value: sha256Sum}
+	*attrs["strip_prefix"] = &bzl.StringExpr{Value: stripPrefix}
+	*attrs["urls"] = &bzl.ListExpr{
+		List: []bzl.Expr{
+			&bzl.StringExpr{Value: mirrorURL},
+			&bzl.StringExpr{Value: ghURL},
+		},
+		ForceMultiLine: true,
+	}
+	urlsKwarg.Before = []bzl.Comment{{Token: "# " + urlComment}}
+
+	return nil
+}
+
+// parseUpgradeDepDirective parses a '# releaser:upgrade-dep org repo' directive
+// and returns the organization and repository name or an error if the directive
+// was not found or malformed.
+func parseUpgradeDepDirective(call *bzl.CallExpr) (orgName, repoName string, err error) {
+	for _, c := range call.Comment().Before {
```

Comment:
I think right here you could `switch` on known things like `go_repository` or `http_archive` and handle those automatically.

Reply:
Added a TODO.